### PR TITLE
EN-1962 defensive measure against duplicate glob pattern

### PR DESCRIPTION
**Pull request by smoy**
What's changed?
* make sure the function play nice with directory ends with / or not
* make sure the glob patterns final results is handled with a set
* support Pathlib.path whenever possible the pydantic models

How'd to test?
* expand unit test on directory ends with "/' or not
* expand unit test on top level yaml (like config file)

### CODE REVIEW

File: iambic/core/utils.py [hunk 1]
```diff
@@ -135,6 +135,9 @@ async def file_regex_search(file_path: str, re_pattern: str) -> Union[str, None]
 
 
 async def gather_templates(repo_dir: str, template_type: str = None) -> list[str]:
+    if repo_dir.endswith("/") and repo_dir != "//":
```
Comment (mdaue2) [inline iambic/core/utils.py:138]: Consider using pathlib to resolve the path, I think pathlib has a bunch of built-ins. The reason for the suggestion is that we also have to account for Windows UNCs, which differ.

File: iambic/core/utils.py [hunk 2]
```diff
@@ -144,14 +147,17 @@ async def gather_templates(repo_dir: str, template_type: str = None) -> list[str
...
+    # to suppress any duplicate for defensive measure
     # Support both yaml and yml extensions for templates
-    file_paths = glob.glob(f"{repo_dir}/**/*.yaml", recursive=True)
-    file_paths += glob.glob(f"{repo_dir}*.yaml", recursive=True)
-    file_paths += glob.glob(f"{repo_dir}/**/*.yml", recursive=True)
-    file_paths += glob.glob(f"{repo_dir}*.yml", recursive=True)
+    file_path_set = set()
+    # >>> glob.glob('**/*.txt', recursive=True)
+    # ['2.txt', 'sub/3.txt']
+    file_path_set.update(glob.glob(f"{repo_dir}/**/*.yaml", recursive=True))
```
Comment (mdaue2) [inline iambic/core/utils.py:156]: Again, would recommend using pathlib.

In pathlib you can do something like: list(p.glob('**/*.yaml')) or p.match('ya?ml'), which would match against the glob style pattern. https://docs.python.org/3/library/pathlib.html#pathlib.PurePath.match

File: iambic/core/utils.py [hunk 3]
```diff
@@ -144,14 +147,17 @@ async def gather_templates(repo_dir: str, template_type: str = None) -> list[str
...
+    file_path_set.update(glob.glob(f"{repo_dir}/**/*.yml", recursive=True))
 
     file_paths = await gather_limit(
-        *[file_regex_search(fp, regex_pattern) for fp in file_paths],
+        *[file_regex_search(fp, regex_pattern) for fp in file_path_set],
```
Comment (mdaue2) [inline iambic/core/utils.py:160]: See comment above using pathlib.match instead

**Review by mdaue2 — DISMISSED** [review]

File: iambic/core/utils.py (hunk 3 above)
Comment (smoy) [reply iambic/core/utils.py:160]: At this point, file_regex_search is actually looking inside the file contents, so I don't think we will use pathlib.match

**Comment by mdaue2** [comment]
On the functional tests, it looks like you may have disabled running all functional tests? Or was this a previous merge? In any case, can you make sure that all functional tests run? I ran unit tests: :heavy_check_mark: 
 and tried to run functional tests, only two ran..

**Review by mdaue2 — APPROVED** [review]

**Pull request merged**